apps/cache/cache.go

```diff
@@ -31,9 +33,9 @@ type Serializer interface {
 type ExportReplace interface {
 	// Replace replaces the cache with what is in external storage.
 	// key is the suggested key which can be used for partioning the cache
-	Replace(cache Unmarshaler, key string)
+	Replace(ctx context.Context, cache Unmarshaler, key string)
```

Comment:
@chlowell @element-of-surprise - is there a way is Go to evolve an interface without breaking change?

In .NET / Java there are 3 ways to do this: 

1. Define a new interface `ExportRelace2` which inherits from `ExportReplace` and defines `Replace(ctx, cache, key)`. Any existing method returning `ExportReplace` now returns `ExportReplace2` - this is not a breaking change. Sometimes app developer has to cast from ExportReplace to ExportReplace2

Reply:
You can't change the interface, that change is breaking.  You can offer a superset:

```go
type ExportReplaceCtx interface {
  ExportReplace

  ReplaceCtx(ctx context.Context, cache Unmarshaler, key string)
  ExportCtx(ctx context.Context, cache Marshaler, key string)
```

```go
if s, ok := b.pmanager.(cache.Serializer); ok {
  suggestedCacheKey := authParams.CacheKey(silent.IsAppCache)
  // If the implementation supports our new ExportReplaceCtx.
  if cache, ok := b.cacheAccessor.(ExportReplaceCtx); ok {
    cache.ReplaceCtx(ctx, s, suggestedCacheKey)
    defer cache.ExportCtx(ctx, s, suggestedCacheKey)
    cache.ReplaceCtx(ctx, s, suggestedCacheKey)
    defer cache.ExportCtx(ctx, s, suggestedCacheKey)
  }else{  // Use the old way.
    b.cacheAccessor.Replace(s, suggestedCacheKey)
    defer b.cacheAccessor.Export(s, suggestedCacheKey)
    b.cacheAccessor.Replace(ctx, s, suggestedCacheKey)
    defer b.cacheAccessor.Export(ctx, s, suggestedCacheKey)
}
```

Reply:
Right, I forgot that method overloading is not available in Go, by design.